File: java/src/br/com/brunohpaiva/osprocessscheduling/algorithms/Fifo.java
package br.com.brunohpaiva.osprocessscheduling.algorithms;

import br.com.brunohpaiva.osprocessscheduling.comparators.ArrivalTimeComparator;
import br.com.brunohpaiva.osprocessscheduling.models.Process;

import java.util.Comparator;
import java.util.List;

public class Fifo {

    public static Double calculate(List<Process> processes, Comparator<Process> tiebreakerComparator) {
        processes.sort(new ArrivalTimeComparator(tiebreakerComparator));

        int currentExecutionTime = 0;
        double totalWaitingTime = 0.0;

        for (Process process : processes) {
            int executedAt = currentExecutionTime;
            currentExecutionTime += process.getBurstTime();
            int waitingTime = executedAt - process.getArrivalTime();
            totalWaitingTime += waitingTime;
        }

        return totalWaitingTime / processes.size();
    }

}

File: typescript/src/algorithms/fifo.ts
import { AlgorithmFunction } from "../types";
import Process from "../models/Process";
import createArrivalTimeComparator from "../comparators/arrivalTime";

const fifo: AlgorithmFunction = (processes, options = {}): number => {
  const sortedProcesses = processes
    .map((process): Process => process.clone())
    .sort(createArrivalTimeComparator(options.tiebreakerComparator));

  let currentExecutionTime = 0;
  let totalWaitingTime = 0;
  for (let index = 0; index < sortedProcesses.length; index++) {
    const process = sortedProcesses[index];

    let executedAt = currentExecutionTime;
    currentExecutionTime += process.burstTime;
    let waitingTime = executedAt - process.arrivalTime;
    totalWaitingTime += waitingTime;

    process.finished = true;
  }

  return totalWaitingTime / sortedProcesses.length;
};

export default fifo;
